sign/sign.go

```diff
@@ -31,12 +32,13 @@ var (
 	ErrNoNamespaceField  = errors.New("transaction must contain namespace field")
 	ErrNoSignatureField  = errors.New("transaction must contain signature field")
 	ErrNoBodyField       = errors.New("transaction must contain body field")
+	ErrNoCreatedField    = errors.New("transaction must contain created field")
 )
 
 type Transaction struct {
 	PersonaTag string          `json:"personaTag"`
 	Namespace  string          `json:"namespace"`
-	Nonce      uint64          `json:"nonce"`
+	Created    int64           `json:"created"`   // unix microsecond timestamp
```

Comment:
Just call it this `Timestamp` so it's self explanatory. I thought we also agree to not use unix microsecond and just use regular unix timestamp?

Reply:
ah, I thought you were referring to the precision of the cache, not the timestamp itself. I made it so that cache operates on seconds. [](edited-with-Axolo)

Reply:
if we are going to just use regular unix timestamp, that should be fine and we can drop that field down to 32 bits [](edited-with-Axolo)

Reply:
And I can certainly rename it Timestamp [](edited-with-Axolo)

Reply:
Hmm... so on further investigation, the golang time.Unix() is inherently a 64 bit value, so we can't drop it down to int32 [](edited-with-Axolo)

Reply:
@smsunarto Tests starts to fail when we switch to Unix second timestamps instead of microsecond timestamp, because some of them assume that even with no other changes to data or payload, each time you sign you get a unique signature. Since we replaced nonce with timestamp, that's only true for messages signed at least one second apart. [](edited-with-Axolo)

Reply:
This uniqueness of signing seems like a good idea that we should preserve [](edited-with-Axolo)

Reply:
That seems to expose a flaw in the design then. I don't think we should be relying on timestamp precision for uniqueness since it's runtime dependent. Javascript/browser-runtime, for example, only have millisecond precision. [](edited-with-Axolo)

Reply:
for now what I'm doing is isolating the nature of that timestamp to the sign package, using sign.TimestampNow() and sign.TimestampAt(time) to generate them. So it will be easy to change.

I can also just start with UnixMilli() instead of UnixMicro(). [](edited-with-Axolo)

Reply:
I'll have that checked in shortly and you can take a look and see if we want to do something further. [](edited-with-Axolo)

Reply:
I say we add an optional salt field to allow distinguishing the messages.  [](edited-with-Axolo)

Reply:
The more I looked into this the more I think we need to understand the problem space better. I'll write up a doc and we can tackle this in a new task.

Reply:
I don't think this should be a different task since we are blocked by this anyway -- we wouldn't be able to cut a release [](edited-with-Axolo)

Reply:
Okay, no problem. Do you mind looking at <https://linear.app/arguslabs/issue/WORLD-1225/guarantee-message-signing-uniqueness-in-sign-package>? I had my thoughts in there about how we might approach this with something like the salt you suggested. We can make it a subtask of the WORLD-1190. [](edited-with-Axolo)

Reply:
replied [](edited-with-Axolo)